specs/extension_image.md

```diff
@@ -34,9 +34,103 @@ sysext images extend `/usr/` (OS vendor tree) and/or `/opt/` (third-party vendor
+for instance, if a sysext extends `/usr` but not `/opt`, `/opt` will remain mutable.
+However, immutability affects the _full_ base directory.
+An extension shipping a single custom path e.g. below `/etc/appconfig-extra/some/sub/path` will still
+render the entirety of `/etc` immutable.
+The base directory or directories remain immutable for as long as the extensions are merged.
+Original state is regained only after all extensions overlaying a base directory have been un-merged.
+
+By implication, merging a confext will result in `/etc` becoming read-only, and merging a sysext might
+render `/usr` and/or `/opt` read-only, depending on the sysext's contents.
+
+### Optional Mutability
+While overlaid base directories are immutable by default, implementations may provide options
+for retaining mutability.
+Retaining mutability may for instance be useful for compatibility with general purpose applications,
+enabling users to operate a "mixed mode" with both system / configuration extensions
+and traditional applications.
+Mixed mode may also be integrated by distributions to facilitate a smooth transition from
+traditional package management to a purely image-based composition of the root file system.
+
+### Extension Overlay [Im]Mutability Modes
+System and configuration extensions may operate in one of three modes.
+
+1. *Immutable mode* - The overlaid base directory is rendered immutable.
+    This is the default.
+2. *Mutable mode* - Writes are directed to an _upperdir_ specified by the user / operator
+    (see "Mutability Mode Configuration" below).
+    This _upperdir_ will contain all changes made to the overlaid base directory.
+    1. _Upperdir_ may be specified to _be_ the base directory: may be used to retain mutability
+       of the base directory after extensions have been merged.
+    2. Alternatively, _upperdir_ may be an entirely separate directory: modifications will be captured
+       but the base directory will remain unchanged, retaining its state from before the extension was merged.
+3. *Ephemeral Mode* - Similar to mutable mode (2.) above but writes are only stored temporarily while
+    extensions are merged, and discarded as soon as extensions are un-merged.
+    Location of temporary storage is implementation specific.
+    Useful for e.g. development and for one-shot validation operations.
+
+### Mutability Mode Configuration
+Immutable mode is the default.
+If none of the configurations outlined below were specified then extension overlays operate
+in immutable mode and base directories are read-only.
+Implementation of any of the below mutable configurations is optional.
+_If_ mutable modes are supported by an implementation, configuration option 1. below _must_ be supported
+for compatibility across implementations.
+
+Mutability modes may be configured in the following ways:
+1. By creating qualified paths or soft-links below `/var/lib/extension-data/`.
```

Comment:
what's the rationale for this dir? it sounds weirdly redundant with its "-data" suffix. That doesn't tell you anything what this is.

I originally assumed the mutable layer would be just /usr.local/. Here you are suggesting an entirely different dir however. what's your rationale for this?

So I think there might be value in placing this in /var/lib/. My thinking has evolved a bit on this. Specifically: I think we should keep the option open to one day have "systemd-sysext commit" and "systemd-confext commit", which would take the current mutable layer and automatically turn it into an proper extension (i.e. an immutable layer), and then opening a new mutable layer. the operation for that would be very simple, just move/rename the upper layer dir, so that it becomes a lower layer dir, mkdir a new upper layer dir, and replace the overlayfs mount with the new arrangement.

Hence, from that angle there's value in having both the immutable sysexts/confexts and the mutable layers on the same fs, so that the move/rename can safely work.

Right now systemd suggests putting confexts in /var/lib/confexts/ and sysexts into /bar/lib/extensions/. I think we should sooner or later probably move the latter to /var/lib/sysexts/ actually. But maybe taking inspiration from that the mutable layer should be:

/var/lib/confexts.mutable/
/var/lib/sysexts.mutable/

maybe?

opinions?

Reply:
other idea would be to put the mutable layer into /var/lib/confexts/mutable/

But I am not sure how much I like that, because it means there would suddenly be a magic layer whose name doesn't really communicate that it's special.

Maybe /var/lib/confexts/@mutable/? 

@keszybz, @t-lo, @pothos, @bluca opinions?

Reply:
Let's not nest and make the whole thing more complicated. A sibling directory in /var/lib/, explicitly called "mutable" or writable or whatever, sounds best to me.

Reply:
I like how descriptive the `/var/lib/confexts.mutable/etc/` and ` /var/lib/sysexts.mutable/usr/` paths are.

As for the rename, maybe keep it consistent and stick to `/var/lib/extensions.mutable/usr/` for now?

Reply:
The proposed directories are quite self-explanatory, I like it. I like Kai's proposal best as it's quite generic. Will amend the PR.